core/local/local.go

```diff
@@ -23,527 +23,363 @@ package local
-	e.Logger.WithField("paused", paused).Debug("Local: Setting paused")
-	e.pauseLock.Lock()
-	defer e.pauseLock.Unlock()
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // just in case, and to shut up go vet...
 
-	if paused && e.pause == nil {
-		e.pause = make(chan interface{})
-	} else if !paused && e.pause != nil {
-		close(e.pause)
-		e.pause = nil
+	// Run setup() before any schedulers, if it's not disabled
+	if !e.options.NoSetup.Bool {
+		logger.Debug("Running setup()")
+		e.initProgress.Modify(pb.WithConstProgress(1, "setup()"))
+		if err := e.runner.Setup(runCtx, engineOut); err != nil {
+			logger.WithField("error", err).Debug("setup() aborted by error")
+			return err
+		}
 	}
-}
-
-func (e *Executor) GetVUs() int64 {
-	return atomic.LoadInt64(&e.numVUs)
-}
+	e.initProgress.Modify(pb.WithHijack(e.getRunStats))
+
+	runCtxDone := runCtx.Done()
+	runScheduler := func(sched lib.Scheduler) {
+		schedConfig := sched.GetConfig()
+		schedStartTime := schedConfig.GetStartTime()
+		schedLogger := logger.WithFields(logrus.Fields{
+			"scheduler": schedConfig.GetName(),
+			"type":      schedConfig.GetType(),
+			"startTime": schedStartTime,
+		})
+		schedProgress := sched.GetProgress()
+
+		// Check if we have to wait before starting the actual scheduler execution
+		if schedStartTime > 0 {
+			startTime := time.Now()
+			schedProgress.Modify(pb.WithProgress(func() (float64, string) {
+				remWait := (schedStartTime - time.Since(startTime))
+				return 0, fmt.Sprintf("waiting %s", pb.GetFixedLengthDuration(remWait, schedStartTime))
+			}))
+
+			schedLogger.Debugf("Waiting for scheduler start time...")
+			select {
+			case <-runCtxDone:
+				runResults <- nil // no error since scheduler hasn't started yet
+				return
+			case <-time.After(schedStartTime):
+				// continue
+			}
+		}
 
-func (e *Executor) SetVUs(num int64) error {
-	if num < 0 {
-		return errors.New("vu count can't be negative")
+		schedProgress.Modify(pb.WithConstProgress(0, "started"))
+		schedLogger.Debugf("Starting scheduler")
+		err := sched.Run(runCtx, engineOut) // scheduler should handle context cancel itself
+		if err == nil {
+			schedLogger.Debugf("Scheduler finished successfully")
+		} else {
+			schedLogger.WithField("error", err).Errorf("Scheduler error")
+		}
+		runResults <- err
 	}
 
-	if atomic.LoadInt64(&e.numVUs) == num {
-		return nil
+	// Start all schedulers at their particular startTime in a separate goroutine...
+	logger.Debug("Start all schedulers...")
+	for _, sched := range e.schedulers {
+		go runScheduler(sched)
 	}
 
-	e.Logger.WithField("vus", num).Debug("Local: Setting VUs")
-
-	if numVUsMax := atomic.LoadInt64(&e.numVUsMax); num > numVUsMax {
-		return errors.Errorf("can't raise vu count (to %d) above vu cap (%d)", num, numVUsMax)
+	// Wait for all schedulers to finish
+	var firstErr error
+	for range e.schedulers {
+		err := <-runResults
+		if err != nil && firstErr == nil {
+			firstErr = err
```

Comment:
Again .. maybe combine all the errors from the different schedulers ?

Reply:
Same reason here as with the initialization - the first error aborts the script, all of the rest of the errors would likely be caused by that and obscure the actually important error. And to be honest, because of the better execution config validation, this error shouldn't ever happen. If it happens, something has very seriously gone wrong :smile: It's not a `panic()` mostly because of the terrible UX of panics...

Reply:
This should be resolved by https://github.com/loadimpact/k6/pull/1007/commits/a3da2afd1ff4d8592f12040750e69498f5fc015a, but feel free to reopen if you disagree. Instead of a comment, I decided to add a debug message, since it both explains what's happening and can also be useful when you're debugging failures in verbose mode.